proxy/apifilterproxy/filters/auth/rancher_token_validate_filter.go

```diff
@@ -0,0 +1,277 @@
+			}
+
+			if accountID == "Forbidden" {
+				log.Errorf("Forbidden")
+				output.Status = http.StatusForbidden
+				return output, nil
+			}
+			if err != nil {
+				log.Errorf("Error getting the accountid and projectid: %v", err)
+				output.Status = http.StatusNotFound
+				return output, nil
+			}
+		} else {
+			accountID, err = getAccountID(f.rancherURL, tokenValue)
+			if accountID == "Unauthorized" {
+				log.Errorf("Unauthorized")
+				output.Status = http.StatusUnauthorized
+				return output, nil
+			}
+			if err != nil {
+				log.Errorf("Error getting the accountid : %v", err)
+				output.Status = http.StatusNotFound
+				return output, nil
+			}
+		}
+
+		//construct the responseBody
+		var headerBody = make(map[string][]string)
+
+		requestHeader := input.Headers
+		for k, v := range requestHeader {
+			headerBody[k] = v
+		}
+
+		headerBody["X-API-Account-Id"] = []string{accountID}
+		if projectID != "" {
+			headerBody["X-API-Project-Id"] = []string{projectID}
+		}
+
+		output.Headers = headerBody
+		output.Status = http.StatusOK
+
+		log.Debugf("Response <= %v", output)
+	}
+	return output, nil
+}
+
+//get the projectID and accountID from rancher API
+func getAccountAndProject(host string, envid string, token string) (string, string, error) {
+
+	client := &http.Client{}
+	requestURL := host + "/v2-beta/projects/" + envid + "/accounts"
+	log.Debugf("requestURL %v", requestURL)
+	req, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		log.Errorf("Cannot connect to the rancher server. Please check the rancher server URL")
+		return "", "", err
+	}
+	cookie := http.Cookie{Name: "token", Value: token}
+	req.AddCookie(&cookie)
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Errorf("Cannot connect to the rancher server. Please check the rancher server URL")
+		return "", "", err
+	}
+	bodyText, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Errorf("Cannot read the reponse body")
+		return "", "", err
+	}
+	authMessage := AuthorizeData{}
+	err = json.Unmarshal(bodyText, &authMessage)
+	if err != nil {
+		log.Errorf("Cannot extract authorization JSON")
+		return "", "", err
+	}
+	if authMessage.Message == "Unauthorized" {
+		log.Errorf("Unauthorized token")
+		err := errors.New("Unauthorized token")
+		return "Unauthorized", "Unauthorized", err
+	}
+
+	projectid := resp.Header.Get("X-Api-Account-Id")
+	userid := resp.Header.Get("X-Api-User-Id")
+	if projectid == "" || userid == "" {
+		log.Errorf("Cannot get projectid or userid")
+		err := errors.New("Forbidden")
+		return "Forbidden", "Forbidden", err
+
+	}
+	if projectid == userid {
+		log.Errorf("Cannot validate project id")
+		err := errors.New("Cannot validate project id")
+		return "", "", err
+
+	}
+
+	log.Debugf("projectid: %v, userid: %v", projectid, userid)
+
+	return projectid, userid, nil
+}
+
+//get the accountID from rancher API
+func getAccountID(host string, token string) (string, error) {
+
+	client := &http.Client{}
```

Comment:
can we create a common http client for all the requests executed against Rancher? And then just reference it from all the methods that need it.